source/dub/project.d

```diff
@@ -116,6 +117,9 @@ class Project {
 	*/
 	bool hasAllDependencies() const { return m_hasAllDependencies; }
```

Comment:
can we change this to `return missingDependencies.length == 0;` and also make the `fetch` function remove from that instead of having that `hasDependentPackagesFetched` boolean in dub.d? I think it would be a lot more fail-safe for future changes on this

Reply:
Thanks for the feedback. The first part I understand, I will change it. The consequence is, the attribute m_hasAllDependencies can be removed at all.

The second part I am not sure. The whole purpose of hasDependentPackagesFetched is to avoid an endless loop due to the recursive function call. In case there are missing packages but they cannot be retrieved, without the Boolean flag there will be an endless loop. The Boolean flag guarantees that only if at least 1 package were successful fetched then the recursive function call will not lead to an endless loop.

As far as I understand your proposed change, endless loops could occur.

Reply:
fetch doesn't call fetch recursively right? After the fetch call it should then check if there are still dependencies unfetched

Reply:
I am not sure whether the proposed solution is getting quite complex to be prepared for future requirements which might come. The fetch is called within a loop. There might be further entries in "versions", which then be handled after the recursive call correctly or maybe incorrectly. Triggering recursive call to function upgrade might add side effects I cannot judge at the moment.

I have to little knowledge about the dub logic here, what consequences it has.
Are you still convinced, adding the recursive function call within the loop after the fetch is the right solution?
If yes, then I will go on with you proposal. If not, I propose to stay with the boolean flag solution until there are further requirements.

Reply:
@WebFreak001 m_hasAllDependencies  is removed as proposed